### PR TITLE
Verify and publish modules in Move

### PR DESCRIPTION
**Pull request by pause125**
This PR is trying to verify modules in native functions. But I got a runtime error:
```
thread '<unnamed>' panicked at 'extension unknown', /home/joker/.cargo/git/checkouts/move-d59e49550a684528/5597bcf/language/move-vm/runtime/src/native_extensions.rs:32:14
```
The error is triggered where getting the `NativeContextExtensions` [here](https://github.com/pause125/rooch/blob/65c6c016455a6095f31250b2ed0ea1b8432abaa4/moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs#L105) and [here](https://github.com/rooch-network/move/blob/47bb63611844e547fed85c9fe7f66d508caa684a/language/move-vm/runtime/src/native_extensions.rs#L32).

I can't figure out why. Help please @jolestar

### CODE REVIEW

**Review by wow-sven — APPROVED** [review]

**Comment by jolestar** [comment]
It is strange, the native context has added when the session creating. Please rebase the branch with the man branch, I need to do some debug.

**Comment by jolestar** [comment]
@pause125 

If the unit test triggers the error, you need to init the native context in the unit test command.

https://github.com/pause125/rooch/blob/65c6c016455a6095f31250b2ed0ea1b8432abaa4/crates/rooch/src/commands/move_cli/commands/unit_test.rs#L50-L56

File: crates/rooch/src/commands/move_cli/commands/unit_test.rs [hunk 1]
```diff
@@ -88,12 +91,18 @@ impl Test {
     }
 }
 
-static STATEDBSTORE: Lazy<Box<StateDBStore>> =
-    Lazy::new(|| Box::new(MoveOSStore::mock_moveos_store().unwrap().statedb));
+static STATEDBSTORE: Lazy<Box<MoveOSResolverProxy<MoveOSStore>>> = Lazy::new(|| {
```
Comment (baichuan3) [inline crates/rooch/src/commands/move_cli/commands/unit_test.rs:94]: `STATEDBSTORE ` should be `MOVEOSSTORE`

File: examples/publish_modules/sources/publish.move [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        // with account 0x5f5c82f7895b07f2cb8751b2001fb7e80f94845f9915bbd56e112af0f68148c8
+        let module_bytes: vector<u8> = x"a11ceb0b060000000b010006020608030e26043406053a32076c8c0108f8014006b802220ada02050cdf02560db5030200000101010200030c00020400000005000100000600010000070201000008030400010907080108010a09010108010b0a0b0108040605060606010708010002070801060c0106080101030107080001080002070801050107090003070801060c090002060801050106090007636f756e7465720f6163636f756e745f73746f726167650f73746f726167655f636f6e7465787407436f756e7465720e53746f72616765436f6e7465787408696e63726561736509696e6372656173655f04696e69740576616c756511676c6f62616c5f626f72726f775f6d75740e676c6f62616c5f6d6f76655f746f0d676c6f62616c5f626f72726f775f5c82f7895b07f2cb8751b2001fb7e80f94845f9915bbd56e112af0f68148c8000000000000000000000000000000000000000000000000000000000000000205205f5c82f7895b07f2cb8751b2001fb7e80f94845f9915bbd56e112af0f68148c800020108030001040001030b0011010201010000050d0b00070038000c010a01100014060100000000000000160b010f0015020200000001060b000b0106000000000000000012003801020301000001060b000700380210001402000000";
```
Comment (baichuan3) [inline examples/publish_modules/sources/publish.move:25]: how to get module_bytes? When implementing the storage test case, I found it not very convenient to generate the module and resource bytes

File: moveos/moveos/src/vm/moveos_vm.rs [hunk 3]
```diff
@@ -317,6 +321,25 @@ where
...
+    /// Resolve pending init functions request registered via the NativeModuleContext.
+    fn resolve_pending_init_functions(&mut self) -> VMResult<()> {
+        let ctx = self
+            .session
+            .get_native_extensions_mut()
+            .get_mut::<NativeModuleContext>();
+        let init_functions = ctx.init_functions.clone();
+        if !init_functions.is_empty() {
+            println!("Pending init functions: {:?}", init_functions);
+            self.execute_init_modules(init_functions)
```
Comment (baichuan3) [inline moveos/moveos/src/vm/moveos_vm.rs:340]: For the init function of the updated module, there will be problems if it is executed again？

**Comment by jolestar** [comment]
Maybe we can support a new command, `rooch move publish_in_move` to test this feature?

File: moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    for module in &compiled_modules {
+        if *module.self_id().address() != account_address {
+            return Err(PartialVMError::new(StatusCode::VERIFICATION_ERROR)
+                .with_message("Module address not match with signer".to_owned()));
+        }
+        let result = moveos_verifier::verifier::verify_module(module, module_context.resolver);
+        match result {
+            Ok(res) => {
+                if res {
+                    module_context.init_functions.push(module.self_id());
```
Comment (jolestar) [inline moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs:121]: the `verify_modules` function is not only used when publishing modules. Or we make the verify_modules to `friend`, and ensure it is only used in the `publish_modules` function.

File: moveos/moveos/src/vm/moveos_vm.rs (hunk 3 above)
Comment (jolestar) [reply moveos/moveos/src/vm/moveos_vm.rs:340]: It is a problem. We can make this TODO and find a solution for the module init function.

**Review by jolestar — APPROVED** [review]

File: moveos/moveos/src/vm/moveos_vm.rs (hunk 3 above)
Comment (jolestar) [reply moveos/moveos/src/vm/moveos_vm.rs:340]: Publish with the `rooch move publish` command only execute the init function when the module first deploys.

File: moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs (hunk 4 above)
Comment (jolestar) [reply moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs:121]: Or we put the init_function into TxContext in the publish_module function. The verify_module function needs to return if the module contains the init function.

**Review by baichuan3 — APPROVED** [review]

File: moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs (hunk 4 above)
Comment (jolestar) [reply moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs]: > > Or we put the init_function into TxContext in the publish_module function. The verify_module function needs to return if the module contains the init function.
> 
> Store init functions in TxContext.map ? The map use name of the value type as key, which may be not suitable to store init functions as we may dynamicly add more init functions with same types.

We need to define a struct and hold all init functions in a vector fields.

**Comment by jolestar** [comment]
I merge this PR first.

**Pull request merged**